### PR TITLE
nvme: pass both --hostid and --hostnqn to discover and connect commands

### PR DESCRIPTION
**Pull request by derekbit**
Longhorn/longhorn#7190

### CODE REVIEW

File: pkg/nvme/nvmecli.go [hunk 1]
```diff
@@ -213,20 +213,34 @@ func listControllers(executor *commonNs.Executor) ([]CliDevice, error) {
...
-		"-q", hostNQN,
 		"-t", DefaultTransportType,
 		"-a", ip,
 		"-s", port,
 		"-o", "json",
 	}
+	if hostID != "" {
+		opts = append(opts, "-I", hostID)
+	}
+	if hostNQN != "" {
```
Comment (innobead) [inline pkg/nvme/nvmecli.go:241]: `hostNQN` is mandatory, so if it's empty, should return err instead? or we just assume it should not be empty here.

File: pkg/nvme/nvmecli.go [hunk 2]
```diff
@@ -213,20 +213,34 @@ func listControllers(executor *commonNs.Executor) ([]CliDevice, error) {
...
-	hostNQN, err := showHostNQN(executor)
-	if err != nil {
-		return nil, err
+func getHostID(executor *commonNs.Executor) (string, error) {
+	outputStr, err := executor.Execute("cat", []string{"/etc/nvme/hostid"}, types.ExecuteTimeout)
+	if err == nil {
+		return strings.TrimSpace(string(outputStr)), nil
+	}
+
+	outputStr, err = executor.Execute("cat", []string{"/sys/class/dmi/id/product_uuid"}, types.ExecuteTimeout)
```
Comment (DamiaSan) [inline pkg/nvme/nvmecli.go:222]: In this way are you overriding the hostid read from `/etc/nvme/hostid`?

Comment (derekbit) [reply pkg/nvme/nvmecli.go:222]: Yes, I just checked nvme-cli codes. nvme-cli tries to read `/etc/nvme/hostid`, so the file should be existing. Do you have any concern?

File: pkg/nvme/nvmecli.go (hunk 1 above)
Comment (derekbit) [reply pkg/nvme/nvmecli.go:241]: Wait, just checked the man page and nvme-cli codes, both hostid and hostnqn are optional.

```
[-q](https://www.mankier.com/1/nvme-discover#-q) <hostnqn>, [--hostnqn](https://www.mankier.com/1/nvme-discover#--hostnqn)=<hostnqn>
Overrides the default host NQN that identifies the NVMe Host. If this option is not specified, the default is read from /etc/nvme/hostnqn first. If that does not exist, the autogenerated NQN value from the NVMe Host kernel module is used next.
```

File: pkg/nvme/nvmecli.go (hunk 2 above)
Comment (DamiaSan) [reply pkg/nvme/nvmecli.go:222]: No, that's ok

**Review by innobead — APPROVED** [review]

**Pull request merged**